sql/tests: TestRandomSyntaxFuncCommon failed

sql/tests.TestRandomSyntaxFuncCommon [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2903571&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2903571&tab=artifacts#/) on master @ [09be4f72a054cf66ef603b1e4525844c9ef547fc](https://github.com/cockroachdb/cockroach/commits/09be4f72a054cf66ef603b1e4525844c9ef547fc):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1913e99]

```
Stack: 

```
goroutine 5888145 [running]:
github.com/cockroachdb/cockroach/pkg/sql/parser.(*lexer).populateErrorDetails(0xc000d90030)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/lexer.go:230 +0x699
github.com/cockroachdb/cockroach/pkg/sql/parser.(*lexer).setErr(0xc000d90030, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/lexer.go:210 +0x5c
github.com/cockroachdb/cockroach/pkg/sql/parser.setErr(0x5bde980, 0xc000d90030, 0x0, 0x0, 0x8)
	sql-gen.y:40 +0x50
github.com/cockroachdb/cockroach/pkg/sql/parser.(*sqlParserImpl).Parse(0xc000d90098, 0x5bde980, 0xc000d90030, 0x0)
	sql-gen.y:10833 +0x6b4a5
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Parser).parse(0xc000d90000, 0x2, 0xc0060636c0, 0x39, 0xc000db4000, 0x12, 0x20, 0x7d62420, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:188 +0x175
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Parser).parseWithDepth(0xc000d90000, 0x1, 0xc0060636c0, 0x39, 0x7d62420, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:168 +0x1d1
github.com/cockroachdb/cockroach/pkg/sql/parser.Parse(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:237
github.com/cockroachdb/cockroach/pkg/sql/tests_test.verifyFormat(0xc0060636c0, 0x39, 0x32, 0x73)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:57 +0x7c
github.com/cockroachdb/cockroach/pkg/sql/tests_test.(*verifyFormatDB).exec(0xc01fa78840, 0xc01fbb0780, 0x5c36580, 0xc00175df40, 0xc0060636c0, 0x39, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:125 +0x73
github.com/cockroachdb/cockroach/pkg/sql/tests_test.TestRandomSyntaxFuncCommon.func1(0x5c36580, 0xc00175df40, 0xc01fa78840, 0xc01ab5c2e0, 0x5b94f00, 0xc0192cd760)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:355 +0x105
github.com/cockroachdb/cockroach/pkg/sql/tests_test.testRandomSyntax.func3(0x5c36580, 0xc00175df40, 0x1, 0x15dbce2, 0x5c365c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:748 +0xc7
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GroupWorkers.func1(0x5c36580, 0xc00175df40, 0xc012612200, 0x7000000054b5b60)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:175 +0x42
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1(0xc012612200, 0x54b5b60)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:166 +0x3a
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc01837faa0, 0xc01837fb00)
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x59
created by golang.org/x/sync/errgroup.(*Group).Go
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x66
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomSyntaxFuncCommon
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxFuncCommon077978143
    test_log_scope.go:74: use -show-logs to present logs inline
    rsg_test.go:731: 5s of 5m0s: 4416 executions, 60 successful
    rsg_test.go:731: 10s of 5m0s: 8526 executions, 87 successful
    rsg_test.go:731: 15s of 5m0s: 12560 executions, 117 successful
    rsg_test.go:731: 20s of 5m0s: 16578 executions, 151 successful
    rsg_test.go:731: 25s of 5m0s: 20507 executions, 182 successful
    rsg_test.go:731: 30s of 5m0s: 24313 executions, 200 successful
    rsg_test.go:731: 35s of 5m0s: 28236 executions, 228 successful
    rsg_test.go:731: 40s of 5m0s: 32059 executions, 247 successful
    rsg_test.go:731: 45s of 5m0s: 35829 executions, 269 successful
    rsg_test.go:731: 50s of 5m0s: 39605 executions, 294 successful
    rsg_test.go:731: 55s of 5m0s: 43232 executions, 328 successful
    rsg_test.go:731: 1m0s of 5m0s: 46957 executions, 350 successful
    rsg_test.go:731: 1m5s of 5m0s: 50557 executions, 371 successful
    rsg_test.go:731: 1m10s of 5m0s: 54143 executions, 397 successful
    rsg_test.go:731: 1m15s of 5m0s: 57768 executions, 419 successful
    rsg_test.go:731: 1m20s of 5m0s: 61452 executions, 447 successful
    rsg_test.go:731: 1m25s of 5m0s: 65042 executions, 471 successful
    rsg_test.go:731: 1m30s of 5m0s: 68569 executions, 489 successful
    rsg_test.go:731: 1m35s of 5m0s: 72004 executions, 506 successful
    rsg_test.go:731: 1m40s of 5m0s: 75482 executions, 533 successful
    rsg_test.go:731: 1m45s of 5m0s: 79019 executions, 552 successful
    rsg_test.go:731: 1m50s of 5m0s: 82539 executions, 574 successful
    rsg_test.go:731: 1m55s of 5m0s: 86078 executions, 598 successful
    rsg_test.go:731: 2m0s of 5m0s: 89642 executions, 615 successful
    rsg_test.go:731: 2m5s of 5m0s: 93112 executions, 632 successful
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxFuncCommon PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFuncCommon.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/tests.TestRandomSyntaxFuncCommon [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2978384&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2978384&tab=artifacts#/) on master @ [5c0b2ec40f06678c0a09c48bcca2904354765734](https://github.com/cockroachdb/cockroach/commits/5c0b2ec40f06678c0a09c48bcca2904354765734):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x191ef79]

```
Stack: 

```
goroutine 6445975 [running]:
github.com/cockroachdb/cockroach/pkg/sql/parser.(*lexer).populateErrorDetails(0xc010031630)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/lexer.go:230 +0x699
github.com/cockroachdb/cockroach/pkg/sql/parser.(*lexer).setErr(0xc010031630, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/lexer.go:210 +0x5c
github.com/cockroachdb/cockroach/pkg/sql/parser.setErr(0x5ca23a0, 0xc010031630, 0x0, 0x0, 0x8)
	sql-gen.y:40 +0x50
github.com/cockroachdb/cockroach/pkg/sql/parser.(*sqlParserImpl).Parse(0xc010031698, 0x5ca23a0, 0xc010031630, 0x0)
	sql-gen.y:10920 +0x73b70
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Parser).parse(0xc010031600, 0x2, 0xc020ddeb90, 0x48, 0xc01053a000, 0x11, 0x20, 0x7e54d40, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:188 +0x175
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Parser).parseWithDepth(0xc010031600, 0x1, 0xc020ddeb90, 0x48, 0x7e54d40, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:168 +0x1d1
github.com/cockroachdb/cockroach/pkg/sql/parser.Parse(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:237
github.com/cockroachdb/cockroach/pkg/sql/tests_test.verifyFormat(0xc020ddeb90, 0x48, 0x41, 0x73)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:57 +0x7c
github.com/cockroachdb/cockroach/pkg/sql/tests_test.(*verifyFormatDB).exec(0xc027ce0330, 0xc002658a80, 0x5cfa6a0, 0xc032331f00, 0xc020ddeb90, 0x48, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:125 +0x73
github.com/cockroachdb/cockroach/pkg/sql/tests_test.TestRandomSyntaxFuncCommon.func1(0x5cfa6a0, 0xc032331f00, 0xc027ce0330, 0xc00eddc840, 0x5c58300, 0xc01e70fe20)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:355 +0x105
github.com/cockroachdb/cockroach/pkg/sql/tests_test.testRandomSyntax.func3(0x5cfa6a0, 0xc032331f00, 0x6, 0xc002571900, 0x5cfa760)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:748 +0xc7
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GroupWorkers.func1(0x5cfa6a0, 0xc032331f00, 0x0, 0x700000000000000)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:175 +0x42
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1(0xc0244b8380, 0x555fcf8)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:166 +0x3a
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc02826bb30, 0xc02826bc80)
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x59
created by golang.org/x/sync/errgroup.(*Group).Go
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x66
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomSyntaxFuncCommon
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxFuncCommon111863109
    test_log_scope.go:74: use -show-logs to present logs inline
    rsg_test.go:731: 5s of 5m0s: 4073 executions, 58 successful
    rsg_test.go:731: 10s of 5m0s: 7817 executions, 84 successful
    rsg_test.go:731: 15s of 5m0s: 11516 executions, 110 successful
    rsg_test.go:731: 20s of 5m0s: 15176 executions, 141 successful
    rsg_test.go:731: 25s of 5m0s: 18772 executions, 161 successful
    rsg_test.go:731: 30s of 5m0s: 22229 executions, 189 successful
    rsg_test.go:731: 35s of 5m0s: 25649 executions, 213 successful
    rsg_test.go:731: 40s of 5m0s: 29143 executions, 244 successful
    rsg_test.go:731: 45s of 5m0s: 32606 executions, 261 successful
    rsg_test.go:731: 50s of 5m0s: 35969 executions, 289 successful
    rsg_test.go:731: 55s of 5m0s: 39417 executions, 309 successful
    rsg_test.go:731: 1m0s of 5m0s: 42835 executions, 328 successful
    rsg_test.go:731: 1m5s of 5m0s: 46321 executions, 349 successful
    rsg_test.go:731: 1m10s of 5m0s: 49701 executions, 373 successful
    rsg_test.go:731: 1m15s of 5m0s: 53072 executions, 392 successful
    rsg_test.go:731: 1m20s of 5m0s: 56400 executions, 410 successful
    rsg_test.go:731: 1m25s of 5m0s: 59847 executions, 429 successful
    rsg_test.go:731: 1m30s of 5m0s: 63251 executions, 447 successful
    rsg_test.go:731: 1m35s of 5m0s: 66545 executions, 466 successful
    rsg_test.go:731: 1m40s of 5m0s: 69819 executions, 478 successful
    rsg_test.go:731: 1m45s of 5m0s: 73150 executions, 497 successful
    rsg_test.go:731: 1m50s of 5m0s: 76512 executions, 518 successful
    rsg_test.go:731: 1m55s of 5m0s: 79826 executions, 538 successful
    rsg_test.go:731: 2m0s of 5m0s: 83036 executions, 550 successful
    rsg_test.go:731: 2m5s of 5m0s: 86352 executions, 577 successful
    rsg_test.go:731: 2m10s of 5m0s: 89616 executions, 590 successful
    rsg_test.go:731: 2m15s of 5m0s: 92867 executions, 610 successful
    rsg_test.go:731: 2m20s of 5m0s: 96150 executions, 633 successful
    rsg_test.go:731: 2m25s of 5m0s: 99456 executions, 647 successful
    rsg_test.go:731: 2m30s of 5m0s: 102705 executions, 667 successful
    rsg_test.go:731: 2m35s of 5m0s: 105866 executions, 684 successful
    rsg_test.go:731: 2m40s of 5m0s: 109104 executions, 693 successful
    rsg_test.go:731: 2m45s of 5m0s: 112398 executions, 710 successful
    rsg_test.go:731: 2m50s of 5m0s: 115539 executions, 726 successful
    rsg_test.go:731: 2m55s of 5m0s: 118744 executions, 747 successful
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxFuncCommon PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #63942 sql/tests: TestRandomSyntaxFuncCommon failed [C-test-failure O-robot branch-release-20.2]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-queries @rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFuncCommon.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/tests.TestRandomSyntaxFuncCommon [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2985382&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2985382&tab=artifacts#/) on master @ [b706c913c07890d04d6bbec4d9d7cd8fea609b95](https://github.com/cockroachdb/cockroach/commits/b706c913c07890d04d6bbec4d9d7cd8fea609b95):


```
    rsg_test.go:731: 1m15s of 5m0s: 55408 executions, 347 successful
    rsg_test.go:731: 1m20s of 5m0s: 58936 executions, 365 successful
    rsg_test.go:731: 1m25s of 5m0s: 62453 executions, 379 successful
    rsg_test.go:731: 1m30s of 5m0s: 65901 executions, 397 successful
    rsg_test.go:731: 1m35s of 5m0s: 69417 executions, 410 successful
    rsg_test.go:731: 1m40s of 5m0s: 72919 executions, 436 successful
    rsg_test.go:731: 1m45s of 5m0s: 76384 executions, 460 successful
    rsg_test.go:731: 1m50s of 5m0s: 79856 executions, 474 successful
    rsg_test.go:731: 1m55s of 5m0s: 83243 executions, 485 successful
    rsg_test.go:731: 2m0s of 5m0s: 86719 executions, 499 successful
    rsg_test.go:731: 2m5s of 5m0s: 90082 executions, 512 successful
    rsg_test.go:731: 2m10s of 5m0s: 93374 executions, 532 successful
    rsg_test.go:731: 2m15s of 5m0s: 96759 executions, 549 successful
    rsg_test.go:731: 2m20s of 5m0s: 100120 executions, 566 successful
    rsg_test.go:731: 2m25s of 5m0s: 103524 executions, 584 successful
    rsg_test.go:731: 2m30s of 5m0s: 106970 executions, 601 successful
    rsg_test.go:731: 2m35s of 5m0s: 110302 executions, 616 successful
    rsg_test.go:731: 2m40s of 5m0s: 113714 executions, 633 successful
    rsg_test.go:731: 2m45s of 5m0s: 117008 executions, 658 successful
    rsg_test.go:731: 2m50s of 5m0s: 120287 executions, 677 successful
    rsg_test.go:731: 2m55s of 5m0s: 123671 executions, 691 successful
    rsg_test.go:731: 3m0s of 5m0s: 126965 executions, 705 successful
    rsg_test.go:731: 3m5s of 5m0s: 130240 executions, 718 successful
    rsg_test.go:731: 3m10s of 5m0s: 133634 executions, 736 successful
    rsg_test.go:731: 3m15s of 5m0s: 136956 executions, 748 successful
    rsg_test.go:731: 3m20s of 5m0s: 140293 executions, 772 successful
    rsg_test.go:731: 3m25s of 5m0s: 143776 executions, 791 successful
    rsg_test.go:731: 3m30s of 5m0s: 147068 executions, 805 successful
    rsg_test.go:731: 3m35s of 5m0s: 150369 executions, 826 successful
    rsg_test.go:731: 3m40s of 5m0s: 153713 executions, 848 successful
    rsg_test.go:731: 3m45s of 5m0s: 157112 executions, 863 successful
    rsg_test.go:731: 3m50s of 5m0s: 160363 executions, 879 successful
    rsg_test.go:731: 3m55s of 5m0s: 163681 executions, 899 successful
    rsg_test.go:731: 4m0s of 5m0s: 166908 executions, 923 successful
    rsg_test.go:731: 4m5s of 5m0s: 170091 executions, 933 successful
    rsg_test.go:731: 4m10s of 5m0s: 173307 executions, 954 successful
    rsg_test.go:731: 4m15s of 5m0s: 176563 executions, 971 successful
    rsg_test.go:731: 4m20s of 5m0s: 179764 executions, 983 successful
    rsg_test.go:731: 4m25s of 5m0s: 183058 executions, 998 successful
    rsg_test.go:731: 4m30s of 5m0s: 186275 executions, 1014 successful
    rsg_test.go:731: 4m35s of 5m0s: 189506 executions, 1027 successful
    rsg_test.go:731: 4m40s of 5m0s: 192715 executions, 1042 successful
    rsg_test.go:731: 4m45s of 5m0s: 195930 executions, 1056 successful
    rsg_test.go:731: 4m50s of 5m0s: 199130 executions, 1068 successful
    rsg_test.go:731: 4m55s of 5m0s: 202365 executions, 1081 successful
    rsg_test.go:767: 205443 executions, 1097 successful
    rsg_test.go:775: Parse followed by Format is not idempotent: "SELECT SUBSTRING ( OPERATOR ( ~ ) DEFAULT COLLATE ident FOR 'string' )" -> "SELECT substring(~DEFAULT COLLATE ident, 1, 'string')" != "SELECT substring((~DEFAULT) COLLATE ident, 1, 'string')"
    rsg_test.go:357: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxFuncCommon053015717
--- FAIL: TestRandomSyntaxFuncCommon (300.29s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxFuncCommon PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #63942 sql/tests: TestRandomSyntaxFuncCommon failed [C-test-failure O-robot branch-release-20.2]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-queries @rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFuncCommon.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/tests.TestRandomSyntaxFuncCommon [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2988805&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2988805&tab=artifacts#/) on master @ [25e26a5f0361fcfbf71cee86f963b93f806604d8](https://github.com/cockroachdb/cockroach/commits/25e26a5f0361fcfbf71cee86f963b93f806604d8):


```
    rsg_test.go:731: 1m15s of 5m0s: 54386 executions, 385 successful
    rsg_test.go:731: 1m20s of 5m0s: 57791 executions, 411 successful
    rsg_test.go:731: 1m25s of 5m0s: 61087 executions, 426 successful
    rsg_test.go:731: 1m30s of 5m0s: 64403 executions, 447 successful
    rsg_test.go:731: 1m35s of 5m0s: 67750 executions, 468 successful
    rsg_test.go:731: 1m40s of 5m0s: 71130 executions, 482 successful
    rsg_test.go:731: 1m45s of 5m0s: 74478 executions, 502 successful
    rsg_test.go:731: 1m50s of 5m0s: 77819 executions, 521 successful
    rsg_test.go:731: 1m55s of 5m0s: 81149 executions, 548 successful
    rsg_test.go:731: 2m0s of 5m0s: 84527 executions, 560 successful
    rsg_test.go:731: 2m5s of 5m0s: 87814 executions, 581 successful
    rsg_test.go:731: 2m10s of 5m0s: 91016 executions, 597 successful
    rsg_test.go:731: 2m15s of 5m0s: 94332 executions, 613 successful
    rsg_test.go:731: 2m20s of 5m0s: 97718 executions, 634 successful
    rsg_test.go:731: 2m25s of 5m0s: 101008 executions, 656 successful
    rsg_test.go:731: 2m30s of 5m0s: 104350 executions, 674 successful
    rsg_test.go:731: 2m35s of 5m0s: 107523 executions, 687 successful
    rsg_test.go:731: 2m40s of 5m0s: 110850 executions, 704 successful
    rsg_test.go:731: 2m45s of 5m0s: 114113 executions, 719 successful
    rsg_test.go:731: 2m50s of 5m0s: 117356 executions, 742 successful
    rsg_test.go:731: 2m55s of 5m0s: 120570 executions, 757 successful
    rsg_test.go:731: 3m0s of 5m0s: 123828 executions, 774 successful
    rsg_test.go:731: 3m5s of 5m0s: 126985 executions, 793 successful
    rsg_test.go:731: 3m10s of 5m0s: 130180 executions, 806 successful
    rsg_test.go:731: 3m15s of 5m0s: 133405 executions, 817 successful
    rsg_test.go:731: 3m20s of 5m0s: 136664 executions, 839 successful
    rsg_test.go:731: 3m25s of 5m0s: 139840 executions, 855 successful
    rsg_test.go:731: 3m30s of 5m0s: 142994 executions, 871 successful
    rsg_test.go:731: 3m35s of 5m0s: 146147 executions, 886 successful
    rsg_test.go:731: 3m40s of 5m0s: 149395 executions, 907 successful
    rsg_test.go:731: 3m45s of 5m0s: 152690 executions, 929 successful
    rsg_test.go:731: 3m50s of 5m0s: 155820 executions, 939 successful
    rsg_test.go:731: 3m55s of 5m0s: 159048 executions, 952 successful
    rsg_test.go:731: 4m0s of 5m0s: 162283 executions, 971 successful
    rsg_test.go:731: 4m5s of 5m0s: 165371 executions, 986 successful
    rsg_test.go:731: 4m10s of 5m0s: 168463 executions, 1003 successful
    rsg_test.go:731: 4m15s of 5m0s: 171700 executions, 1016 successful
    rsg_test.go:731: 4m20s of 5m0s: 174933 executions, 1030 successful
    rsg_test.go:731: 4m25s of 5m0s: 178148 executions, 1047 successful
    rsg_test.go:731: 4m30s of 5m0s: 181235 executions, 1064 successful
    rsg_test.go:731: 4m35s of 5m0s: 184407 executions, 1081 successful
    rsg_test.go:731: 4m40s of 5m0s: 187646 executions, 1097 successful
    rsg_test.go:731: 4m45s of 5m0s: 190760 executions, 1112 successful
    rsg_test.go:731: 4m50s of 5m0s: 193917 executions, 1127 successful
    rsg_test.go:731: 4m55s of 5m0s: 197109 executions, 1136 successful
    rsg_test.go:767: 200141 executions, 1153 successful
    rsg_test.go:775: Parse followed by Format is not idempotent: "SELECT LOCALTIME ( OPERATOR ( ~ ) DEFAULT COLLATE RESET )" -> "SELECT localtime(~DEFAULT COLLATE reset)" != "SELECT localtime((~DEFAULT) COLLATE reset)"
    rsg_test.go:357: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxFuncCommon165154359
--- FAIL: TestRandomSyntaxFuncCommon (300.33s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxFuncCommon PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #63942 sql/tests: TestRandomSyntaxFuncCommon failed [C-test-failure O-robot branch-release-20.2]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-queries @rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFuncCommon.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/tests.TestRandomSyntaxFuncCommon [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3273417&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3273417&tab=artifacts#/) on master @ [cab185ff71f0924953d987fe6ffd14efdd32a3a0](https://github.com/cockroachdb/cockroach/commits/cab185ff71f0924953d987fe6ffd14efdd32a3a0):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x18dadcf]

```
Stack: 

```
goroutine 4364110 [running]:
github.com/cockroachdb/cockroach/pkg/sql/parser.(*lexer).populateErrorDetails(0xc0160e8b30)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/lexer.go:245 +0x6af
github.com/cockroachdb/cockroach/pkg/sql/parser.(*lexer).setErr(0xc0160e8b30, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/lexer.go:225 +0x5c
github.com/cockroachdb/cockroach/pkg/sql/parser.setErr(0x675b628, 0xc0160e8b30, 0x0, 0x0, 0x8)
	sql-gen.y:40 +0x50
github.com/cockroachdb/cockroach/pkg/sql/parser.(*sqlParserImpl).Parse(0xc0160e8b98, 0x675b628, 0xc0160e8b30, 0x0)
	sql-gen.y:11359 +0x72c8e
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Parser).parse(0xc0160e8b00, 0x2, 0xc0173c99a0, 0x94, 0xc012740800, 0x24, 0x40, 0x8b19f00, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:195 +0x175
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Parser).parseWithDepth(0xc0160e8b00, 0x1, 0xc0173c99a0, 0x94, 0x8b19f00, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:175 +0x1d1
github.com/cockroachdb/cockroach/pkg/sql/parser.Parse(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:247
github.com/cockroachdb/cockroach/pkg/sql/tests_test.verifyFormat(0xc0173c99a0, 0x94, 0xc01e835d38, 0x556fbc)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:58 +0x7c
github.com/cockroachdb/cockroach/pkg/sql/tests_test.(*verifyFormatDB).execWithTimeout(0xc01f3e85a0, 0xc013824000, 0x67a49a8, 0xc01445ca40, 0xc0173c99a0, 0x94, 0xdf8475800, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:130 +0x73
github.com/cockroachdb/cockroach/pkg/sql/tests_test.(*verifyFormatDB).exec(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:125
github.com/cockroachdb/cockroach/pkg/sql/tests_test.TestRandomSyntaxFuncCommon.func1(0x67a49a8, 0xc01445ca40, 0xc01f3e85a0, 0xc00988b8e0, 0x670ea00, 0xc015fc3a00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:360 +0x117
github.com/cockroachdb/cockroach/pkg/sql/tests_test.testRandomSyntax.func3(0x67a49a8, 0xc01445ca40, 0x6, 0x12eb713, 0x67a49e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:753 +0xc7
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GroupWorkers.func1(0x67a49a8, 0xc01445ca40, 0xc0142d4380, 0x580f6d0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:175 +0x42
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1(0x0, 0x580f778)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:166 +0x3a
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc02082b230, 0xc02082b380)
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x59
created by golang.org/x/sync/errgroup.(*Group).Go
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x66
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomSyntaxFuncCommon
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxFuncCommon530542064
    test_log_scope.go:74: use -show-logs to present logs inline
    rsg_test.go:736: 5s of 5m0s: 4284 executions, 63 successful
    rsg_test.go:736: 10s of 5m0s: 8149 executions, 91 successful
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxFuncCommon PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #66753 sql/tests: TestRandomSyntaxFuncCommon failed [C-test-failure O-robot branch-release-21.1]
- #63942 sql/tests: TestRandomSyntaxFuncCommon failed [C-test-failure O-robot T-sql-experience branch-release-20.2]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-experience rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFuncCommon.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/tests.TestRandomSyntaxFuncCommon [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3344852&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3344852&tab=artifacts#/) on master @ [46cef2c6f0b36ba2f7d551c8ab017832c1b9d592](https://github.com/cockroachdb/cockroach/commits/46cef2c6f0b36ba2f7d551c8ab017832c1b9d592):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x191fa0f]

```
Stack: 

```
goroutine 4163372 [running]:
github.com/cockroachdb/cockroach/pkg/sql/parser.(*lexer).populateErrorDetails(0xc03637e130)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/lexer.go:261 +0x6af
github.com/cockroachdb/cockroach/pkg/sql/parser.(*lexer).setErr(0xc03637e130, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/lexer.go:241 +0x5c
github.com/cockroachdb/cockroach/pkg/sql/parser.setErr(0x69620c8, 0xc03637e130, 0x0, 0x0, 0x8)
	sql-gen.y:40 +0x50
github.com/cockroachdb/cockroach/pkg/sql/parser.(*sqlParserImpl).Parse(0xc03637e198, 0x69620c8, 0xc03637e130, 0x0)
	sql-gen.y:11428 +0x738ee
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Parser).parse(0xc03637e100, 0x2, 0xc006a58f00, 0x24, 0xc02aa86f00, 0xa, 0x10, 0x8dbd7c0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:195 +0x175
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Parser).parseWithDepth(0xc03637e100, 0x1, 0xc006a58f00, 0x24, 0x8dbd7c0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:175 +0x1d1
github.com/cockroachdb/cockroach/pkg/sql/parser.Parse(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:247
github.com/cockroachdb/cockroach/pkg/sql/tests_test.verifyFormat(0xc006a58f00, 0x24, 0xc0771c9d38, 0x55705c)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:58 +0x7c
github.com/cockroachdb/cockroach/pkg/sql/tests_test.(*verifyFormatDB).execWithTimeout(0xc04c6d2840, 0xc00272de00, 0x69acac8, 0xc019145bc0, 0xc006a58f00, 0x24, 0xdf8475800, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:130 +0x73
github.com/cockroachdb/cockroach/pkg/sql/tests_test.(*verifyFormatDB).exec(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:125
github.com/cockroachdb/cockroach/pkg/sql/tests_test.TestRandomSyntaxFuncCommon.func1(0x69acac8, 0xc019145bc0, 0xc04c6d2840, 0xc0890a3100, 0x6913700, 0xc08fa7bd80)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:360 +0x117
github.com/cockroachdb/cockroach/pkg/sql/tests_test.testRandomSyntax.func3(0x69acac8, 0xc019145bc0, 0x3, 0x131ce73, 0x69acb00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/tests/rsg_test.go:753 +0xc7
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GroupWorkers.func1(0x69acac8, 0xc019145bc0, 0xc0975b8280, 0x59b73a8)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:175 +0x42
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1(0x0, 0x59b7450)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:166 +0x3a
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc00ff31470, 0xc00ff31530)
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x59
created by golang.org/x/sync/errgroup.(*Group).Go
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x66
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomSyntaxFuncCommon
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxFuncCommon718186303
    test_log_scope.go:74: use -show-logs to present logs inline
    rsg_test.go:736: 5s of 5m0s: 4112 executions, 55 successful
    rsg_test.go:736: 10s of 5m0s: 7912 executions, 83 successful
    rsg_test.go:736: 15s of 5m0s: 11599 executions, 111 successful
    rsg_test.go:736: 20s of 5m0s: 15261 executions, 128 successful
    rsg_test.go:736: 25s of 5m0s: 18992 executions, 149 successful
    rsg_test.go:736: 30s of 5m0s: 22583 executions, 173 successful
    rsg_test.go:736: 35s of 5m0s: 26253 executions, 199 successful
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxFuncCommon PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #66753 sql/tests: TestRandomSyntaxFuncCommon failed [C-test-failure O-robot branch-release-21.1]
- #63942 sql/tests: TestRandomSyntaxFuncCommon failed [C-test-failure O-robot T-sql-experience branch-release-20.2]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-experience rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFuncCommon.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/float
+++ b/pkg/sql/logictest/testdata/logic_test/float
@@ -139,3 +139,23 @@ query FFF
 SELECT -0.1234567890123456, 123456789012345.6, 1234567.890123456
 ----
 -0.1234567890123457 123456789012345.7 1234567.890123457
+
+# Regression test to make sure `IS NAN` does not coerce NaN to a string.
+statement error unsupported comparison operator: <tuple> = <decimal>
+SELECT ROW() IS NAN
+
+statement error unsupported comparison operator: <tuple> != <decimal>
+SELECT ROW() IS NOT NAN
+
+statement error unsupported comparison operator: <string> = <decimal>
+SELECT 'NaN'::string IS NAN
+
+query BB
+SELECT 'nan'::float IS NAN, 'nan'::float IS NOT NAN
+----
+true  false
+
+query BB
+SELECT 'nan'::decimal IS NAN, 'nan'::decimal IS NOT NAN
+----
+true  false

--- a/pkg/sql/parser/sql.y
+++ b/pkg/sql/parser/sql.y
@@ -11230,11 +11230,19 @@ a_expr:
   }
 | a_expr IS NAN %prec IS
   {
-    $$.val = &tree.ComparisonExpr{Operator: tree.MakeComparisonOperator(tree.EQ), Left: $1.expr(), Right: tree.NewStrVal("NaN")}
+    $$.val = &tree.ComparisonExpr{
+      Operator: tree.MakeComparisonOperator(tree.EQ),
+      Left: $1.expr(),
+      Right: tree.NewNumVal(constant.MakeFloat64(math.NaN()), "NaN", false /*negative*/),
+    }
   }
 | a_expr IS NOT NAN %prec IS
   {
-    $$.val = &tree.ComparisonExpr{Operator: tree.MakeComparisonOperator(tree.NE), Left: $1.expr(), Right: tree.NewStrVal("NaN")}
+    $$.val = &tree.ComparisonExpr{
+      Operator: tree.MakeComparisonOperator(tree.NE),
+      Left: $1.expr(),
+      Right: tree.NewNumVal(constant.MakeFloat64(math.NaN()), "NaN", false /*negative*/),
+    }
   }
 | a_expr IS NULL %prec IS
   {

--- a/pkg/sql/parser/testdata/select_exprs
+++ b/pkg/sql/parser/testdata/select_exprs
@@ -1662,7 +1662,7 @@ SELECT a IS NAN
 ----
 SELECT a = 'NaN' -- normalized!
 SELECT ((a) = ('NaN')) -- fully parenthesized
-SELECT a = '_' -- literals removed
+SELECT a = _ -- literals removed
 SELECT _ = 'NaN' -- identifiers removed
 
 parse
@@ -1670,7 +1670,7 @@ SELECT a IS NOT NAN
 ----
 SELECT a != 'NaN' -- normalized!
 SELECT ((a) != ('NaN')) -- fully parenthesized
-SELECT a != '_' -- literals removed
+SELECT a != _ -- literals removed
 SELECT _ != 'NaN' -- identifiers removed
 
 parse

--- a/pkg/sql/sem/tree/constant.go
+++ b/pkg/sql/sem/tree/constant.go
@@ -168,6 +168,8 @@ func (expr *NumVal) Format(ctx *FmtCtx) {
 	s := expr.origString
 	if s == "" {
 		s = expr.value.String()
+	} else if strings.EqualFold(s, "NaN") {
+		s = "'NaN'"
 	}
 	if expr.negative {
 		ctx.WriteByte('-')
@@ -305,11 +307,18 @@ func (expr *NumVal) ResolveAsType(
 		}
 		return &expr.resInt, nil
 	case types.FloatFamily:
-		f, _ := constant.Float64Val(expr.value)
-		if expr.negative {
-			f = -f
+		if strings.EqualFold(expr.origString, "NaN") {
+			// We need to check NaN separately since expr.value is unknownVal for NaN.
+			// TODO(sql-experience): unknownVal is also used for +Inf and -Inf,
+			// so we may need to handle those in the future too.
+			expr.resFloat = DFloat(math.NaN())
+		} else {
+			f, _ := constant.Float64Val(expr.value)
+			if expr.negative {
+				f = -f
+			}
+			expr.resFloat = DFloat(f)
 		}
-		expr.resFloat = DFloat(f)
 		return &expr.resFloat, nil
 	case types.DecimalFamily:
 		dd := &expr.resDecimal

